python/ffsim/states/slater.py

```diff
@@ -0,0 +1,296 @@
+        seed: Either a Numpy random generator, an integer seed for the random number
+            generator or ``None``.
+    Returns:
+        A numpy array with the position of the sampled electrons.
+    """
+    rng = np.random.default_rng(seed)
+
+    position = [rng.choice(norb, p=np.real(np.diag(rdm)) / nelec)]
+    marginal = [np.real(np.diag(rdm) / nelec)[position[0]]]
+
+    for k in range(nelec - 1):
+        pos_array = np.array(position, dtype=int)
+
+        empty_orbitals = np.setdiff1d(np.arange(norb), pos_array, assume_unique=True)
+
+        u_conditionals, marginals = _generate_conditionals_unormalized(
+            rdm, pos_array, empty_orbitals, marginal[-1]
+        )
+
+        conditionals = np.real(u_conditionals)
+        conditionals /= np.sum(conditionals)
+
+        index = rng.choice(len(empty_orbitals), p=conditionals)
+
+        new_pos = empty_orbitals[index]
+
+        position.append(new_pos)
+        position.sort()
+        marginal.append(marginals[index])
+    return np.array(position, dtype=int)
+
+
+def _positions_to_bit_array(positions: np.ndarray, norb: int) -> np.ndarray:
+    """Transforms electronic configurations defined by the position of the electrons
+    to the occupation representation.
+
+    Args:
+        positions: A Numpy array of positions of electrons. Each row corresponds
+            to an electronic configuration.
+        norb: number of orbitals.
+
+    Returns:
+        A 2Dimensional Numpy array of electronic configurations represented by
+        the orbital occupancy. Each row is an electronic configuration.
+    """
+
+    n_samples = positions.shape[0]
+    fock = np.zeros((n_samples, norb), dtype=int)
+    for i in range(n_samples):
+        fock[i, norb - 1 - positions[i]] = 1
+    return fock
+
+
+def _sample_spinless_direct(
+    rdm: np.ndarray,
+    shots: int,
+    seed: np.random.Generator | int | None = None,
+) -> np.ndarray:
+    """Collect samples of electronic configurations from a Slater determinant for
+    spin-polarized systems.
+
+    The Slater determinat is defined by its one-body reduced density matrix (RDM).
+    The sampler uses a determinantal point process to auto-regressively produce
+    uncorrelated samples.
+
+    Args:
+        rdm: The one-body reduced density matrix that defines the Slater determinant
+            This is either a single Numpy array specifying the 1-RDM of a
+            spin-polarized system, or a pair of Numpy arrays where each element
+            of the pair contains the 1-RDM for each spin sector.
+        shots: Number of samples to collect.
+        seed: Either a Numpy random generator, an integer seed for the random number
+            generator or ``None``.
+
+    Returns:
+        A 2D Numpy array with samples of electronic configurations.
+        Each row is a sample.
+    """
+
+    rng = np.random.default_rng(seed)
+    norb, _ = rdm.shape
+    nelec = round(np.real(np.sum(np.diag(rdm))))
```

Comment:
Instead of computing `nelec`, have this function take it as input. Then, we can handle the `np.zeros` and `np.ones` cases here, based on the value of `nelec`. This will make the code in the main function much cleaner.